text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+breaking changes.
+
+### Lifecycle Hooks
+
+Legacy components have 13 major lifecycle hooks that run during 4 major phases
+of the component lifecycle, with some hooks running during multiple phases:
+
+1. **Initialization:**
+   * `init`
+
+2. **Initial Render:**
+   * `willInsertElement`
+   * `didInsertElement`,
+   * `didReceiveAttrs`
+   * `willRender`
+   * `didRender`.
+
+3. **Rerenders and Updates:**
+   * `didReceiveAttrs`
+   * `didUpdateAttrs`
+   * `willUpdate`
+   * `didUpdate`
+   * `willRender`
+   * `didRender`
+
+4. **Destruction:**
+   * `willDestroyElement`
+   * `didDestroyElement`
+   * `willDestroy`
+   * `destroy`
+
+
+Many of these hooks have overlapping or redundant functionality, and it's fairly
+confusing when to use which and what the differences are. We can simplify this
+cycle in a number of ways:
+
+* Hooks that run during multiple phases such as `didRender` and
+  `didRecieveAttrs` can be convenient at times, but also add mental overhead and
+  redundancy. We can remove these in favor of clearly delineated hooks which
+  only run during _one_ phase.
+
+* "Bookend" methods (`did*` and `will*`) can be confusing, since they require
+  some specific knowledge of what the "bookended" functionality is. For
+  instance, users almost _always_ want to use `didInsertElement` and
+  `willDestroyElement`, but the existence of their opposite bookends can make
+  this confusing. Additionally, the fact that `didReceiveAttrs` and
+  `didUpdateAttrs` do _not_ have opposing bookends is inconsistent with this
+  pattern.
+
+* Hooks that run immediately _before_ a render such as `didUpdateAttrs` are not
+  particularly useful, because the only functionality they have which
+  post-render methods do not is to modify a bound value without triggering a
+  rerender:
+
+  ```js
+  willUpdateAttrs() {
+    if (this.someBoolArg) {
+      this.set('someBoundValue', true);
+    }
+  }
+  ```
+
+  These types of values can almost always be better expressed using computed
+  properties, which will update when accessed during render, preventing
+  rerenders. Conversely, much imperative logic _must_ be done after a render
+  (manipulating an element directly, for instance), so post render hooks are
+  the most sensible.
+
+* The element based methods can be removed in favor of element modifiers,
+  which are discussed later on.
+
+Based on these considerations, we can reduce these hooks to one per major phase
+of the component's lifecycle:
+
+The **`constructor`** can be used for initial setup of the component. This
+effectively replaces `init`, and allows users to setup state before _any_
+renders occur.
+
+**`didCreate`** runs after the first full render of the component, once the
+component has been fully created. It is closest to `didInsertElement` in legacy
+components, but has been renamed since components no longer have an `element`
+property.
+
+**`didUpdate`** runs after the component has been updated, including all changes
```

Comment:
what does update mean exactly, does it mean the  values are updated, and not rendered, or does it mean rendered, or does it mean all rendering through to its descends have completed? The RFC here may want to really lock this down.

Reply:
Is this invoked depth first, or breadth first, top -> down, or bottom -> up?

Reply:
I see, bellow you do elaborate on the ordering. Should section then link to that, or maybe hint that it will be addressed?

Reply:
I think we can address it here. `didUpdate` does mix some concerns at the moment, it is invoked in two distinct situations:

* `args` were changed, whether or not this caused a rerender
* A rerender occured.

In both cases, it occurs _immediately_ after the rerender has finished for this component, so all child rerenders/`didUpdate`s have finished (at least in my experimentation). I don't think we can/should specify ordering relative to siblings, but I think spelling this out would be good (we can do the same for `didCreate`)